drivers/spi/spi_pl022.c

```diff
@@ -0,0 +1,541 @@
+	cr0 |= (op & SPI_MODE_CPHA) ? SSP_CR0_MASK_SPH : 0;
+
+	cr1 = 0;
+	cr1 |= SSP_CR1_MASK_SSE; /* Always enable SPI */
+	cr1 |= (op & SPI_MODE_LOOP) ? SSP_CR1_MASK_LBM : 0;
+
+	SSP_WRITE_REG(SSP_CPSR(cfg->reg), prescale);
+	SSP_WRITE_REG(SSP_CR0(cfg->reg), cr0);
+	SSP_WRITE_REG(SSP_CR1(cfg->reg), cr1);
+
+	data->ctx.config = spicfg;
+
+	return 0;
+}
+
+/*
+ * calculate actual transfer length.
+ * If one buffer is invalid, use length of others,
+ * IF both buffer is valid, using length of shorter one.
+ */
+static inline size_t spi_pl022_xfer_length(const void *txbuf, const size_t txlen,
+					   void *rxbuf, const size_t rxlen)
+{
+	size_t min = (txlen < rxlen ? txlen : rxlen);
+
+	if (!txbuf && txlen == 0) {
+		return rxlen;
+	} else if (!rxbuf && rxlen == 0) {
+		return txlen;
+	}
+	return min;
+}
+
+static size_t spi_pl022_xfer(const struct device *dev, const void *txbuf, const size_t txlen,
+			     void *rxbuf, const size_t rxlen)
+{
+	const struct spi_pl022_cfg *cfg = dev->config;
+	const size_t xferlen = spi_pl022_xfer_length(txbuf, txlen, rxbuf, rxlen);
+	size_t rx_remaining = xferlen;
+	size_t tx_remaining = xferlen;
+	uint32_t txd;
+	uint32_t rxd;
+	size_t idx;
+
+	/* Ensure writable */
+	while (!SSP_TX_FIFO_EMPTY(cfg->reg))
+		;
+	/* Drain RX FIFO */
+	while (SSP_RX_FIFO_NOT_EMPTY(cfg->reg))
+		SSP_READ_REG(SSP_DR(cfg->reg));
+
+	while (rx_remaining > 0 || tx_remaining > 0) {
+		/* Fill up fifo with available TX data */
+		while (SSP_TX_FIFO_NOT_FULL(cfg->reg) && tx_remaining > 0) {
+			/* Send 0 in the case of read operation */
+			txd = 0;
+			idx = xferlen - tx_remaining;
+
+			if (txbuf) {
+				txd = ((uint8_t *)txbuf)[idx];
+			}
+			SSP_WRITE_REG(SSP_DR(cfg->reg), txd);
+			--tx_remaining;
+		}
+		while (SSP_RX_FIFO_NOT_EMPTY(cfg->reg) && rx_remaining > 0) {
+			/* Discard received data if buffer not assigned */
+			rxd = SSP_READ_REG(SSP_DR(cfg->reg));
+			idx = xferlen - rx_remaining;
+
+			if (rxbuf) {
+				((uint8_t *)rxbuf)[idx] = (uint8_t)rxd;
+			}
+			--rx_remaining;
+		}
+	}
+
+	return xferlen;
+}
+
+static inline bool spi_pl022_transfer_ongoing(struct spi_pl022_data *data)
+{
+	return spi_context_tx_on(&data->ctx) ||
+	       spi_context_rx_on(&data->ctx);
+}
+
+/* API Functions */
+
+static int spi_pl022_transceive(const struct device *dev,
+				const struct spi_config *config,
+				const struct spi_buf_set *tx_bufs,
+				const struct spi_buf_set *rx_bufs)
+{
+	struct spi_pl022_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int ret;
+
+	spi_context_lock(&data->ctx, false, NULL, config);
+
+	ret = spi_pl022_configure(dev, config);
+	if (ret < 0) {
+		goto error;
+	}
+
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(ctx, true);
+
+	do {
+		size_t len = spi_pl022_xfer(dev, ctx->tx_buf, ctx->tx_len,
+					    ctx->rx_buf, ctx->rx_len);
+
+		spi_context_update_tx(ctx, 1, len);
+		spi_context_update_rx(ctx, 1, len);
+
+	} while (spi_pl022_transfer_ongoing(data));
```

Comment:
Is there any possibility to get this driver working through interrupts instead of polling?

Reply:
> Is there any possibility to get this driver working through interrupts instead of polling?

Currently, only work with polling. Not handles interrupt.
But I think it should also support the CONFIG_SPI_ASYNC option by handling interrupts
I have a plan to implement an interrupts handler later.